streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java

```diff
@@ -631,19 +629,9 @@ private void checkBrokerVersionCompatibility() throws StreamsException {
 
     }
 
-    private void validateStartOnce() {
-        try {
-            if (setState(RUNNING)) {
-                return;
-            }
-        } catch (StreamsException e) {
-            // do nothing, will throw
-        }
-        throw new IllegalStateException("Cannot start again.");
-    }
-
     /**
      * Start the {@code KafkaStreams} instance by starting all its threads.
+     * This function is expected to be called only once during the life cycle of the client.
```

Comment:
There's no real reasons; only because currently we create the threads in constructor's and the threads cannot be restarted:

```
     * It is never legal to start a thread more than once.
     * In particular, a thread may not be restarted once it has completed
     * execution.
```

We can consider lifting this restrictions by creating new threads in the `start()` call instead of in constructor in the future.